optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -186,20 +186,87 @@
+    /**
+     * Runs all tuples of the stream through a given @{@link UniConstraintCollector} and converts them into a new
+     * {@link UniConstraintStream} which only has a single tuple, the result of applying {@link UniConstraintCollector}.
+     *
+     * @param collector never null, the collector to perform the grouping operation with
+     * @param <ResultContainer_> the mutable accumulation type (often hidden as an implementation detail)
+     * @param <Result_> the type of a fact in the destination {@link UniConstraintStream}'s tuple
+     * @return never null
+     */
+    <ResultContainer_, Result_> UniConstraintStream<Result_> groupBy(
             UniConstraintCollector<A, ResultContainer_, Result_> collector);
 
-    <GroupKey_> UniConstraintStream<GroupKey_> groupBy(
-            Function<A, GroupKey_> groupKeyMapping);
-
+    /**
+     * Convert the {@link UniConstraintStream} to a different {@link UniConstraintStream}, containing of the set of
+     * tuples resulting from applying the group key mapping function on all tuples of the original stream.
+     * Neither tuple of the new stream will {@link Objects#equals(Object, Object)} any other.
+     *
+     * @param groupKeyMapping never null, mapping function to convert each element in the stream to a different element
+     * @param <GroupKey_> the type of a fact in the destination {@link UniConstraintStream}'s tuple
+     * @return never null
+     */
+    <GroupKey_> UniConstraintStream<GroupKey_> groupBy(Function<A, GroupKey_> groupKeyMapping);
+
+    /**
+     * Expands the {@link UniConstraintStream} to a {@link BiConstraintStream}, consisting of tuples which have:
+     *
+     * <ul>
+     *     <li>As the first fact, the value resulting from applying the group key mapping function on the fact from the
+     *     original tuple.
+     *     Neither such facts of the new stream will {@link Objects#equals(Object, Object)} any other.</li>
+     *     <li>As the second fact, the value of applying a given {@link UniConstraintCollector} on all the facts of the
+     *     original tuple with a matching group key.</li>
+     * </ul>
+     *
+     * @param groupKeyMapping never null, function to convert a fact in original tuple to a different fact
+     * @param <GroupKey_> the type of the first fact in the destination {@link BiConstraintStream}'s tuple
+     * @param <ResultContainer_> the mutable accumulation type (often hidden as an implementation detail)
+     * @param <Result_> the type of the second fact in the destination {@link BiConstraintStream}'s tuple
+     * @return never null
+     */
     <GroupKey_, ResultContainer_, Result_> BiConstraintStream<GroupKey_, Result_> groupBy(
             Function<A, GroupKey_> groupKeyMapping,
             UniConstraintCollector<A, ResultContainer_, Result_> collector);
 
+    /**
+     * Expand the {@link UniConstraintStream} to a {@link BiConstraintStream}, consisting of tuples which have:
```

Comment:
I am not convinced about using "Expand" in this context.
Yes, the cardinality increases.
But the element count decreases.

When we turn a Uni<Apple> into a Bi<Airplane, Orange>, it doesn't feel as expanding.
Better example: We turn a Uni<Person> into a Bi<Age, Gender> and that person stream has [Ann(40, female), Beth(40, female), Carl (31, male)] that becomes [{40, female}, {31, male}]. Does that feel like an expansion?

Reply:
Instead of "expand", I 'd just use "Convert".
I'd also mention that the new stream will have less or equal elements as the original.